Suggest `Self::CONST` when trying to use an associated constant lexically

### Code

```Rust
struct Foo;
// const A_CONST: usize = 1;
impl Foo {
    const A_CONST: usize = 1;
    fn foo() -> usize {
        // const A_CONST: usize = 1;
        A_CONST
    }
}
```


### Current output

```Shell
error[E0425]: cannot find value `A_CONST` in this scope
 --> src/lib.rs:7:9
  |
7 |         A_CONST
  |         ^^^^^^^ not found in this scope
```


### Desired output

```Shell
help: try referring to the associated constant: `Self::A_CONST`
```


### Rationale and extra context

Associated constants work entirely differently from constants at a higher scope (module) or lower (function) which are looked up lexically — if either of the other two declarations in the sample code are uncommented, compilation succeeds. This may be surprising to beginners ([example that inspired this post](https://stackoverflow.com/questions/75199496/why-cant-you-put-constants-in-an-impl-block)), and the “what they meant” is fairly obvious, so this should be a good opportunity to help in the error message.

Additionally, if the suggestion is machine-applicable then it can be used to help refactor code when a constant was intentionally converted into an associated constant.

### Other cases

_No response_

### Anything else?

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"euclio"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

Further precedent: the compiler already provides this help for calling associated functions.

```rust
struct Foo;
impl Foo {
    fn bar() {}
    fn foo() {
        bar();
    }
}
```

```
error[E0425]: cannot find function `bar` in this scope
 --> src/lib.rs:5:9
  |
5 |         bar();
  |         ^^^ not found in this scope
  |
help: consider using the associated function
  |
5 |         Self::bar();
  |         ~~~~~~~~~
```

--- a/compiler/rustc_resolve/src/late/diagnostics.rs
+++ b/compiler/rustc_resolve/src/late/diagnostics.rs
@@ -227,20 +227,27 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
                     && let Some(FnCtxt::Assoc(_)) = fn_kind.ctxt()
                     && let Some(items) = self.diagnostic_metadata.current_impl_items
                     && let Some(item) = items.iter().find(|i| {
-                        if let AssocItemKind::Fn(_) = &i.kind && i.ident.name == item_str.name
+                        if let AssocItemKind::Fn(..) | AssocItemKind::Const(..) = &i.kind
+                            && i.ident.name == item_str.name
                         {
                             debug!(?item_str.name);
                             return true
                         }
                         false
                     })
-                    && let AssocItemKind::Fn(fn_) = &item.kind
                 {
-                    debug!(?fn_);
-                    let self_sugg = if fn_.sig.decl.has_self() { "self." } else { "Self::" };
+                    let self_sugg = match &item.kind {
+                        AssocItemKind::Fn(fn_) if fn_.sig.decl.has_self() => "self.",
+                        _ => "Self::",
+                    };
+
                     Some((
                         item_span.shrink_to_lo(),
-                        "consider using the associated function",
+                        match &item.kind {
+                            AssocItemKind::Fn(..) => "consider using the associated function",
+                            AssocItemKind::Const(..) => "consider using the associated constant",
+                            _ => unreachable!("item kind was filtered above"),
+                        },
                         self_sugg.to_string()
                     ))
                 } else {

--- a/tests/ui/suggestions/assoc-const-without-self.rs
+++ b/tests/ui/suggestions/assoc-const-without-self.rs
@@ -1,0 +1,11 @@
+struct Foo;
+
+impl Foo {
+    const A_CONST: usize = 1;
+
+    fn foo() -> usize {
+        A_CONST //~ ERROR cannot find value `A_CONST` in this scope
+    }
+}
+
+fn main() {}

--- a/tests/ui/suggestions/assoc-const-without-self.stderr
+++ b/tests/ui/suggestions/assoc-const-without-self.stderr
@@ -1,0 +1,14 @@
+error[E0425]: cannot find value `A_CONST` in this scope
+  --> $DIR/assoc-const-without-self.rs:7:9
+   |
+LL |         A_CONST
+   |         ^^^^^^^ not found in this scope
+   |
+help: consider using the associated constant
+   |
+LL |         Self::A_CONST
+   |         ++++++
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0425`.

